QuizApp/Network/NetworkService.swift

```diff
@@ -0,0 +1,220 @@
+            }
+            
+            guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode)
+                else {
+                    DispatchQueue.main.async {
+                        completionHandler(nil, .serverError)
+                    }
+                    return
+                    
+            }
+            
+            guard let data = data else {
+                DispatchQueue.main.async {
+                    completionHandler(nil, .noDataError)
+                }
+                return
+            }
+            
+            guard let value = try? JSONDecoder().decode(QuizzesResult.self, from: data) else {
+                DispatchQueue.main.async {
+                    completionHandler(nil, .decodingError)
+                }
+                return
+            }
+            
+            let quizzes = value.quizzes
+            
+            DispatchQueue.main.async {
+                completionHandler(quizzes, nil)
+            }
+        }
+        
+        task.resume()
+    }
+    
+    func executeQuizResultsUploadRequest(_ request: URLRequest, bodyData: Data, completionHandler: @escaping (QuizResultsUploadResponse?, RequestError?) -> Void) {
+        let task = URLSession.shared.uploadTask(with: request, from: bodyData) {
+            data, response, err in
+            guard err == nil else {
+                DispatchQueue.main.async {
+                    completionHandler(nil, .clientError)
+                }
+                return
+            }
+            
+            //var serverResponse: ServerResponse
+            
+            //            guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode)
+            //                else {
+            //                    DispatchQueue.main.async {
+            //                        completionHandler(.ok, .serverError)
+            //
+            //                    }
+            //                    return
+            //
+            //            }
+            
+            let httpResponse = response as? HTTPURLResponse
+            if httpResponse != nil {
+                var serverResponse: ServerResponse?
+                var requestError: RequestError?
+                if (200...299).contains(httpResponse!.statusCode) {
+                    serverResponse = .ok
+                    requestError = nil
+                } else {
+                    requestError = .serverError
+                    switch(httpResponse!.statusCode) {
+                    case 400:
+                        serverResponse = .badRequest
+                    case 401:
+                        serverResponse = .unauthorized
+                    case 403:
+                        serverResponse = .forbidden
+                    case 404:
+                        serverResponse = .notFound
+                    case _:
+                        serverResponse = nil
```

Comment:
Even if you don't match the cases you still want to know the error code. You could create an additional case inside ServerResponse called 
`case unknownResponse(Int)` with an associated Int value.
Then here you would have a default case where you would catch any unexpected error code.
```
default:
    serverResponse = .unknownResponse(statusCode)
```